build.gradle

```diff
@@ -47,35 +49,35 @@ subprojects {
 
 
   dependencies {
-    compile 'org.slf4j:slf4j-api:1.7.32'
+    api 'org.slf4j:slf4j-api:1.7.32'
 
-    compile "com.fasterxml.jackson.core:jackson-annotations:${jacksonVersion}"
-    compile "com.fasterxml.jackson.core:jackson-databind:${jacksonVersion}"
-    compile "com.fasterxml.jackson.datatype:jackson-datatype-jsr310:${jacksonVersion}"
+    api "com.fasterxml.jackson.core:jackson-annotations:${jacksonVersion}"
+    api "com.fasterxml.jackson.core:jackson-databind:${jacksonVersion}"
+    api "com.fasterxml.jackson.datatype:jackson-datatype-jsr310:${jacksonVersion}"
 
-    compile 'com.google.guava:guava:31.0.1-jre'
+    api 'com.google.guava:guava:31.0.1-jre'
 
-    compile "org.reflections:reflections:${reflectionsVersion}"
-    compile 'com.esotericsoftware.kryo:kryo5:5.1.1'
+    api "org.reflections:reflections:${reflectionsVersion}"
+    api 'com.esotericsoftware.kryo:kryo5:5.2.1'
 
-    compile('commons-io:commons-io:2.11.0'){
+    api('commons-io:commons-io:2.11.0'){
       exclude group: 'commons-logging', module: 'commons-logging'
     }
-    compile 'org.apache.commons:commons-text:1.9'
-    compile 'org.apache.commons:commons-lang3:3.12.0'
+    api 'org.apache.commons:commons-text:1.9'
+    api 'org.apache.commons:commons-lang3:3.12.0'
 
     //java 10 no longer has these included by default, so explicit dependency is needed.
-    compile 'javax.xml.bind:jaxb-api:2.3.1'
-    compile 'com.sun.xml.bind:jaxb-core:2.3.0.1'
-    runtime 'com.sun.xml.bind:jaxb-impl:2.3.3'
-    compile 'javax.activation:activation:1.1.1'
+    api 'javax.xml.bind:jaxb-api:2.3.1'
+    api 'com.sun.xml.bind:jaxb-core:2.3.0.1'
+    runtimeOnly 'com.sun.xml.bind:jaxb-impl:2.3.3'
+    api 'javax.activation:activation:1.1.1'
 
-    implementation 'org.threeten:threeten-extra:1.7.0'
+    api 'org.threeten:threeten-extra:1.7.0'
 
-    testCompile 'junit:junit:4.+'
-    testCompile 'org.slf4j:slf4j-simple:1.7.32'
+    testImplementation 'junit:junit:4.+'
+    testImplementation 'org.slf4j:slf4j-simple:1.7.32'
 
-    compile 'com.github.zafarkhaja:java-semver:0.9.0'
+    api 'com.github.zafarkhaja:java-semver:0.9.0'
```

Comment:
Just checking: can some/most of these `api` keywords actually be `implementation` instead? If I remember correctly this is about boxing certain dependencies so that implementing applications cannot use the dependencies (or something), which sounds better than just leaking all dependencies.

Reply:
It's tricky. If you publish it to maven, it will leak them anyway, since api will be set to 'compile' and implementation to 'runtime' in the resulting pom.xml.

Setting them all to 'api' is the backwards compatible option, since it is exactly the same as 'compile'. I think it is best to later change this, if we want to.

if we do, most of these are actually API and not implementation: jaxb, jackson, antlr, semver, etc are all required in some level of api. The following libraries could at least be moved to implementation:

- kryo
- apache commons
- guava

But it could very well break other projects if we do so, unless they do complicated classloader trickery to separate these dependencies from their own. On the other hand, those will not be hard to fix.

Reply:
Fair enough, this is definitely the most safe option then. 

I would argue that breaking other projects makes sense during a major version bump, and boxing these dependencies helps with accidental incorrect imports/auto complete clutter etc.